src/librustc_trait_selection/traits/error_reporting/suggestions.rs

```diff
@@ -147,6 +147,126 @@ crate trait InferCtxtExt<'tcx> {
     fn suggest_new_overflow_limit(&self, err: &mut DiagnosticBuilder<'_>);
 }
 
+fn predicate_constraint(generics: &hir::Generics<'_>, pred: String) -> (Span, String) {
+    (
+        generics.where_clause.span_for_predicates_or_empty_place().shrink_to_hi(),
+        format!(
+            "{} {} ",
+            if !generics.where_clause.predicates.is_empty() { "," } else { " where" },
+            pred,
+        ),
+    )
+}
+
+/// Type parameter needs more bounds. The trivial case is `T` `where T: Bound`, but
+/// it can also be an `impl Trait` param that needs to be decomposed to a type
+/// param for cleaner code.
+fn suggest_restriction(
+    generics: &hir::Generics<'_>,
+    msg: &str,
+    err: &mut DiagnosticBuilder<'_>,
+    fn_sig: Option<&hir::FnSig<'_>>,
+    projection: Option<&ty::ProjectionTy<'_>>,
+    trait_ref: &ty::PolyTraitRef<'_>,
+) {
+    let span = generics.where_clause.span_for_predicates_or_empty_place();
+    if !span.from_expansion() && span.desugaring_kind().is_none() {
+        // Given `fn foo(t: impl Trait)` where `Trait` requires assoc type `A`...
+        if let Some((name, fn_sig)) = fn_sig.and_then(|sig| {
+            projection.and_then(|p| {
+                // Shenanigans to get the `Trait` from the `impl Trait`.
+                match p.self_ty().kind {
+                    ty::Param(param) => {
+                        // `fn foo(t: impl Trait)`
+                        //                 ^^^^^ get this string
+                        param
+                            .name
+                            .as_str()
+                            .strip_prefix("impl")
+                            .map(|s| (s.trim_start().to_string(), sig))
+                    }
```

Comment:
Actually, looked at the whole diff, you have the `hir::Generics`, and you could either use that (tricky) or you could pass in the `predicates_of` for the item. Which, I think if you `instantiate_identity`, you get a flat list of all of the predicates in scope, and they should still have `Span`s.

You can presumably search for `trait_ref` in the list of predicate and get the corresponding `Span`.
Which should be *exactly* the `Foo` in `impl Foo + Bar`, that corresponds to the trait that's missing an associated type.

Reply:
Bonus: you can probably use the `Span`s in `predicates_of` for other suggestions elsewhere - their initial purpose was automatically applicable suggestions for `type` aliases not being WF, back in late 2018, before that stuff fell through the f loor.

Reply:
So I added span labels for each `TraitPredicate`, and this is what I got:

```
error[E0277]: `<impl Iterator as std::iter::Iterator>::Item` doesn't implement `std::fmt::Debug`
  --> src/test/ui/suggestions/impl-trait-with-missing-bounds.rs:6:13
   |
4  | fn foo(constraints: impl  Iterator) {
   |                     --------------
   |                     |     |
   |                     |     Binder(TraitPredicate(<impl Iterator as std::iter::Iterator>)) NotConst
   |                     Binder(TraitPredicate(<impl Iterator as std::marker::Sized>)) NotConst
5  |     for constraint in constraints {
6  |         qux(constraint);
   |             ^^^^^^^^^^ `<impl Iterator as std::iter::Iterator>::Item` cannot be formatted using `{:?}` because it doesn't implement `std::fmt::Debug`
...
27 | fn qux(_: impl std::fmt::Debug) {}
   |    ---         --------------- required by this bound in `qux`
   |
   = help: the trait `std::fmt::Debug` is not implemented for `<impl Iterator as std::iter::Iterator>::Item`
help: introduce a type parameter with a trait bound instead of using `impl Trait`
   |
4  | fn foo<T: Iterator>(constraints: T) where <T as std::iter::Iterator>::Item: std::fmt::Debug  {
   |       ^^^^^^^^^^^^^              ^  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

I can see how it could be useful for some things, but the problem I see is that because this is an assoc type bound instead of setting it a type we cannot use the sytnax `impl Iterator<Item = Something>`. These spans would be great for that, but in this case we need to sadly rely only on extending the `where` clause to `<X as Iterator>::Item: Trait`.

Reply:
You have two options: switching to `impl Iterator<Item = T>` where `T` is a new type parameter and adding a `T: Trait` bound, or `impl Iterator<Item: Trait>`, although I'm not sure the latter works yet or when it will work and it's probably still unstable anyway.

Anyway I was commenting only on the approach of getting the trait bound from the `ty::Param` "name": my point is that you can get its *definition* and work with that, which lets you get the *real snippet* where `impl Foo+Bar` was written (and each `Foo` and `Bar` *in that* if you look in the `predicates_of`).

Reply:
> `impl Iterator<Item: Trait>`

Yeah, that's not valid today.

> Anyway I was commenting only on the approach

I understand.

> which lets you get the _real snippet_ where `impl Foo+Bar` was written

The current output for a `impl Foo + Bar` case is the following:

```
error[E0277]: `<impl Iterator + std::fmt::Debug as std::iter::Iterator>::Item` doesn't implement `std::fmt::Debug`
  --> src/test/ui/suggestions/impl-trait-with-missing-bounds.rs:37:13
   |
37 |         qux(constraint);
   |             ^^^^^^^^^^ `<impl Iterator + std::fmt::Debug as std::iter::Iterator>::Item` cannot be formatted using `{:?}` because it doesn't implement `std::fmt::Debug`
...
42 | fn qux(_: impl std::fmt::Debug) {}
   |    ---         --------------- required by this bound in `qux`
   |
   = help: the trait `std::fmt::Debug` is not implemented for `<impl Iterator + std::fmt::Debug as std::iter::Iterator>::Item`
help: introduce a type parameter with a trait bound instead of using `impl Trait`
   |
35 | fn bak<T: Iterator + std::fmt::Debug>(constraints: T) where <T as std::iter::Iterator>::Item: std::fmt::Debug  {
   |       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^              ^  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

We need the whole `Foo + Bar` for the type param `T: Foo + Bar`. The only place where having the span to `Foo` for its snippet could be useful as things stand now is for the suggestion message where we now use `<impl Foo + Bar as Foo>::AssocType` where we could say `<_ as Foo>::AssocType` instead, but I feel that leaving it as it is is reasonable.

I think suggesting `fn foo<T: Debug>(constraints: impl Iterator<Item = T>) {}` is reasonable, but at that point the only thing left is the tension between ease of implementation of the suggestion and stylistic preference. The later seems to be more stylistically preferable, but don't know what the code to suggest it will look like yet.